src/kubernetes/reflector.rs

```diff
@@ -24,26 +24,53 @@ pub async fn custom_reflector<K, W>(
     let mut delay_queue = DelayQueue::default();
     loop {
         tokio::select! {
-            Some(Ok(event)) = stream.next() => {
-                match event {
-                    watcher::Event::Applied(_) => {
-                        // Immediately apply event
-                        store.apply_watcher_event(&event);
-                    }
-                    watcher::Event::Deleted(_) => {
-                        // Delay reconciling any `Deleted` events
-                        delay_queue.insert(event.to_owned(), delay_deletion);
-                    }
-                    watcher::Event::Restarted(_) => {
-                        // Clear all delayed events when the cache is refreshed
-                        delay_queue.clear();
-                        store.apply_watcher_event(&event);
-                    }
+            result = stream.next() => {
+                match result {
+                    Some(Ok(event)) => {
+                        match event {
+                            // Immediately reoncile `Applied` event
+                            watcher::Event::Applied(_) => {
+                                trace!("Processing Applied event.");
+                                store.apply_watcher_event(&event);
+                            }
+                            // Delay reconciling any `Deleted` events
+                            watcher::Event::Deleted(_) => {
+                                trace!("Queuing Deleted event.");
+                                delay_queue.insert(event.to_owned(), delay_deletion);
+                            }
+                            // Clear all delayed events on `Restarted` events
+                            watcher::Event::Restarted(_) => {
+                                trace!("Processing Restarted event.");
+                                delay_queue.clear();
+                                store.apply_watcher_event(&event);
+                            }
+                        }
+                    },
+                    Some(Err(error)) => {
+                        warn!(message = "Watcher stream got an error.", error = ?error);
+                    },
+                    None => {
+                        info!("Watcher stream has terminated.");
+                        break;
+                    },
```

Comment:
Theoretically this should never happen because the watcher stream never yields `None`:

https://docs.rs/kube-runtime/0.71.0/src/kube_runtime/watcher.rs.html#234-237

I'd be ok with doing `unreachable("watcher stream never ends")`.

Otherwise do you know what would happen if this `break` is hit? Does Vector shutdown? Something else?

Reply:
Vector shouldn't stop, it would stop looping here and stop updating our cache - this was another "correctness" arm, and I can see `unreachable` as acceptable as well. Deferring to y'all here, interested in what you think @lukesteensen.

Reply:
Gotcha, yeah, I think since the implementation never actually returns `None` I'd prefer to mark it `unreachable`. If they start returning `None` there then we would need to deal with it by either:

* Stopping Vector
* Restarting the watcher

I think simply ceasing to update the cache will confuse users.

Reply:
I guess it depends if we consider never returning `None` part of the API contract or just an implementation detail. I could imagine an API where these streams are expected to last as long as the underlying resource and then finish, but I don't know enough about the subject to really say.

Reply:
Yeah, [`watch_object`](https://docs.rs/kube-runtime/0.71.0/src/kube_runtime/watcher.rs.html#247-272) behaves that way (returns `None` when the object is deleted). Not returning `None` isn't explicitly mentioned in the function documentation for `watcher`. We could guard against it, but I think that would mean doing one of the options I mentioned above: either shutting down Vector or restarting the watcher.